### PR TITLE
Add fonction to workflow

### PR DESCRIPTION
**Pull request by itkg-nanne**

### CODE REVIEW

File: WorkflowAdminBundle/Controller/Api/FonctionController.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace OpenOrchestra\WorkflowAdminBundle\Controller\Api;
+use OpenOrchestra\BaseApi\Facade\FacadeInterface;
```
Comment (itkg-ngilain) [inline WorkflowAdminBundle/Controller/Api/FonctionController.php]: a little newline here maybe

File: WorkflowAdminBundle/DependencyInjection/OpenOrchestraWorkflowAdminExtension.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+class OpenOrchestraWorkflowAdminExtension extends Extension
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
+        $loader->load('form.yml');
+        $loader->load('leftpanel.yml');
```
Comment (itkg-ngilain) [inline WorkflowAdminBundle/DependencyInjection/OpenOrchestraWorkflowAdminExtension.php]: can we supress the term leftPanel and replace it with navigationPanel ? I find this naming innacurate
Same as rightPanel which should mentionned as blockPanel on my mind

Comment (nicolasThal) [inline WorkflowAdminBundle/DependencyInjection/OpenOrchestraWorkflowAdminExtension.php]: I think that this change will have a little more consequence, you will also have to change the LeftPanelInterface.

We can add a refacto ticket for this change

Comment (itkg-ngilain) [inline WorkflowAdminBundle/DependencyInjection/OpenOrchestraWorkflowAdminExtension.php]: I'm totally for making a general refacto, I'll add that ticket.

My point was to stop using this wrong terminology when we're creating new files and can avoid it, in order to simplify this future refacto and initiate the move :)

**Comment by itkg-nanne** [comment]
first rename work done @nicolasThal

File: GroupBundle/DataFixtures/MongoDB/LoadGroupData.php [hunk 3]
```diff
@@ -10,6 +10,7 @@
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\GeneralNodesPanelStrategy;
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\TreeNodesPanelStrategy;
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\TreeTemplatePanelStrategy;
+use OpenOrchestra\WorkflowAdminBundle\LeftPanel\Strategies\FonctionPanelStrategy;
```
Comment (nicolasThal) [inline GroupBundle/DataFixtures/MongoDB/LoadGroupData.php]: I think you forgot to change this namespace :)

File: WorkflowFonctionAdminBundle/Controller/Admin/WorkflowFonctionController.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Class WorkflowFonctionController
+ */
+class WorkflowFonctionController extends AbstractAdminController
+{
+    /**
+     * @param Request $request
+     * @param string  $workflowFonctionId
+     *
+     * @Config\Route("/workflowfonction/form/{workflowFonctionId}", name="open_orchestra_backoffice_workflowfonction_form")
```
Comment (nicolasThal) [inline WorkflowFonctionAdminBundle/Controller/Admin/WorkflowFonctionController.php]: you can put workflowfonction as a prefix for all the route of the class, See RoleController for an exemple

File: WorkflowFonctionAdminBundle/Controller/Api/WorkflowFonctionController.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace OpenOrchestra\WorkflowFonctionAdminBundle\Controller\Api;
+use OpenOrchestra\BaseApi\Facade\FacadeInterface;
```
Comment (nicolasThal) [inline WorkflowFonctionAdminBundle/Controller/Api/WorkflowFonctionController.php]: empty line missing

File: WorkflowFonctionAdminBundle/Form/Type/WorkflowFonctionType.php [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Returns the name of this type.
+     *
+     * @return string The name of this type
+     */
+    public function getName()
+    {
+        return 'workflowfonction';
+    }
+
```
Comment (nicolasThal) [inline WorkflowFonctionAdminBundle/Form/Type/WorkflowFonctionType.php]: empty line to delete

File: WorkflowFonctionAdminBundle/LeftPanel/Strategies/WorkflowFonctionPanelStrategy.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+     * @var WorkflowFonctionRepositoryInterface
+     */
+    protected $workflowFonctionRepository;
+
+    /**
+     * @param WorkflowFonctionRepositoryInterface $workflowFonctionRepository
+     */
+    public function __construct(WorkflowFonctionRepositoryInterface $workflowFonctionRepository)
+    {
+        $this->workflowFonctionRepository = $workflowFonctionRepository;
```
Comment (nicolasThal) [inline WorkflowFonctionAdminBundle/LeftPanel/Strategies/WorkflowFonctionPanelStrategy.php]: If you don't need the repository, why did you inject it?

**Comment by nicolasThal** [comment]
As a general remark, I would avoid the usage of the string workflowfonction, I would rather use workflow_fonction.

In the routing, you can use workflow-fonction

File: WorkflowFonctionAdminBundle/LeftPanel/Strategies/WorkflowFonctionPanelStrategy.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return self::ADMINISTRATION;
+    }
+
+    /**
+     * @return string
+     */
+    public function getName()
+    {
+        return 'workflowfonctions';
```
Comment (nicolasThal) [inline WorkflowFonctionAdminBundle/LeftPanel/Strategies/WorkflowFonctionPanelStrategy.php]: I would rather put workflow_function only

**Comment by itkg-nanne** [comment]
renamme done @nicolasThal

File: GroupBundle/DataFixtures/MongoDB/LoadGroupData.php [hunk 9]
```diff
@@ -10,6 +10,7 @@
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\GeneralNodesPanelStrategy;
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\TreeNodesPanelStrategy;
 use OpenOrchestra\Backoffice\LeftPanel\Strategies\TreeTemplatePanelStrategy;
+use OpenOrchestra\WorkflowFonctionAdminBundle\LeftPanel\Strategies\WorkflowFonctionPanelStrategy;
```
Comment (nicolasThal) [inline GroupBundle/DataFixtures/MongoDB/LoadGroupData.php]: I think that this namespace is still not the right one :)

File: GroupBundle/DataFixtures/MongoDB/LoadGroupData.php [hunk 10]
```diff
@@ -71,6 +72,7 @@ protected function generateGroup($name, $siteNumber, $referenceName, $role = nul
             $group->addRole('ROLE_FROM_DRAFT_TO_PENDING');
             $group->addRole('ROLE_FROM_PENDING_TO_PUBLISHED');
             $group->addRole('ROLE_FROM_PUBLISHED_TO_DRAFT');
+            $group->addRole(WorkflowFonctionPanelStrategy::ROLE_ACCESS_WORKFLOWFONCTION);
```
Comment (nicolasThal) [inline GroupBundle/DataFixtures/MongoDB/LoadGroupData.php]: same here

**Comment by nicolasThal** [comment]
:+1: except the small mistake in the fixtures :)

File: WorkflowFunctionAdminBundle/Controller/Admin/WorkflowFunctionController.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * @Config\Route("workflow-function")
+ */
+class WorkflowFunctionController extends AbstractAdminController
+{
+    /**
+     * @param Request $request
+     * @param string  $workflowFunctionId
+     *
+     * @Config\Route("/form/{workflowFunctionId}", name="open_orchestra_backoffice_workflow_function_form")
```
Comment (itkg-ngilain) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowFunctionController.php]: form or edit ?

Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowFunctionController.php]: form as for all controller in BackofficeBundle/Controller/*

Comment (itkg-ngilain) [inline WorkflowFunctionAdminBundle/Controller/Admin/WorkflowFunctionController.php]: Ok so

File: WorkflowFunctionAdminBundle/OpenOrchestraWorkflowFunctionAdminBundle.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+class OpenOrchestraWorkflowFunctionAdminBundle extends Bundle
+{
+    /**
+     * @param ContainerBuilder $container
+     */
+    public function build(ContainerBuilder $container)
+    {
+        parent::build($container);
+    }
+}
```
Comment (itkg-ngilain) [inline WorkflowFunctionAdminBundle/OpenOrchestraWorkflowFunctionAdminBundle.php]: newline at EOF

File: WorkflowFunctionAdminBundle/Controller/Api/WorkflowFunctionController.php [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        $workflowFunctionCollection = $this->get('open_orchestra_workflow_function.repository.workflow_function')->findAll();
+
+        return $this->get('open_orchestra_api.transformer_manager')->get('workflow_function_collection')->transform($workflowFunctionCollection);
+    }
+
+    /**
+     * @param string $workflowFunctionId
+     *
+     * @Config\Route("/{workflowFunctionId}/delete", name="open_orchestra_api_workflow_function_delete")
```
Comment (itkg-ngilain) [inline WorkflowFunctionAdminBundle/Controller/Api/WorkflowFunctionController.php]: We get /form/{workflowFunctionId} and /{workflowFunctionId}/delete, maybe we could unify the routing logic

Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Controller/Api/WorkflowFunctionController.php]: I agree, I would suggesst /{workflowFunctionId}/delete and /{workflowFunctionId}/form

Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Controller/Api/WorkflowFunctionController.php]: same as for form or edit. This logic is used in all others controller. Maybe refacto?

**Comment by nicolasThal** [comment]
:+1:

**Pull request merged**